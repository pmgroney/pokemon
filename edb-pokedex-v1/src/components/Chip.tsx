import { ReactNode } from 'react';

type Props = {
  children: ReactNode;
};

export const Chip = ({ children }: Props) => {
  return (
    <div className="flex flex-wrap justify-center space-x-2">
      <span className="px-4 py-2 rounded-full text-gray-500 bg-gray-200 font-semibold text-sm flex align-center w-max cursor-pointer active:bg-gray-300 transition duration-300 ease">
        {children}
      </span>
    </div>
  );
};
