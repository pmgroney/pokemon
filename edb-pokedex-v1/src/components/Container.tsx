import { ReactNode } from 'react';

export const Container = ({ children }: { children: ReactNode }) => {
  return <div className="container max-w-5xl mx-auto">{children}</div>;
};
