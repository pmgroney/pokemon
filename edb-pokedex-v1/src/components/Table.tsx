import { SerializedPokemon } from '@/types';
import { fadeInUp } from '@/utils/animations';
import { motion } from 'framer-motion';
import Link from 'next/link';
import React, { useMemo } from 'react';
import { useTable, useSortBy, Column } from 'react-table';

type Props = {
  data: SerializedPokemon[];
};

const columns: Array<Column<SerializedPokemon>> = [
  {
    Header: '  ',
    accessor: (data: any) => (
      <Link key={data.id} href={`/pokemon/${data.name}`} data-testid={`data-${data.id}`}>
        <motion.div variants={fadeInUp} whileHover={{ scale: 1.03 }} whileTap={{ scale: 1 }}>
          <div data-testid={`data-${data.id}-name`} className="text-l font-bold ml-10">
            {data.name}
          </div>
        </motion.div>
      </Link>
    ),
  },
  {
    Header: ' ',
    accessor: (data: any) => (
      <img
        src={data.image}
        alt={data.name}
        width="35"
        className="ml-0 rounded-lg border-slate-300 border p-1 bg-white h-9 drop-shadow-2xl"
      />
    ),
  },
  {
    Header: 'HP',
    accessor: (data: any) => data.stats['HP'],
  },
  {
    Header: 'Attack',
    accessor: (data: any) => data.stats['Attack'],
  },
  {
    Header: 'Defense',
    accessor: (data: any) => data.stats['Defense'],
  },
  {
    Header: 'Sp. Attack',
    accessor: (data: any) => data.stats['Sp. Attack'],
  },
  {
    Header: 'Sp. Defense',
    accessor: (data: any) => data.stats['Sp. Defense'],
  },
  {
    Header: 'Speed',
    accessor: (data: any) => data.stats['Speed'],
  },
];

const Table = (props: Props) => {
  const data = useMemo(() => props.data, [props.data]);
  const { getTableBodyProps, headerGroups, rows, prepareRow } = useTable(
    {
      columns,
      data,
    },
    useSortBy,
  );

  return (
    <div>
      <table className="w-full">
        <thead className="text-left">
          {headerGroups.map((headerGroup) => (
            <tr {...headerGroup.getHeaderGroupProps()} key={headerGroup.id}>
              {headerGroup.headers.map((column) => (
                <th
                  {...column.getHeaderProps(column.getSortByToggleProps())}
                  key={column.id}
                  scope="col"
                >
                  {column.render('Header')}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map((row) => {
            prepareRow(row);
            return (
              <tr {...row.getRowProps()} key={row.id}>
                {row.cells.map((cell) => {
                  return <td {...cell.getCellProps()}>{cell.render('Cell')} </td>;
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
};

export default Table;
