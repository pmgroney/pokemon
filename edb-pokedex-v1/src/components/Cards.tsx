import { Chip } from '@/components/Chip';
import { SerializedPokemon } from '@/types';
import { fadeInUp } from '@/utils/animations';
import { motion } from 'framer-motion';
import Link from 'next/link';

export type Props = { data: SerializedPokemon[] };

export const Cards = ({ data }: Props) => {
  return (
    <div className="flex flex-wrap gap-10 justify-between">
      {data.map((pokemon) => (
        <Link
          key={pokemon.id}
          href={`/pokemon/${pokemon.name}`}
          data-testid={`pokemon-${pokemon.id}`}
        >
          <motion.div
            variants={fadeInUp}
            whileHover={{ scale: 1.03 }}
            whileTap={{ scale: 1 }}
            className="rounded-lg bg-slate-100 border-slate-300 border p-5 max-w-md cursor-pointer flex-none w-6/12"
          >
            <div>
              {/* Name */}
              <div data-testid={`pokemon-${pokemon.id}-name`} className="text-2xl">
                {pokemon.name}
              </div>

              <div className="flex items-center">
                {/* Types */}
                <div className="flex gap-1 flex-col items-start">
                  {pokemon.type?.map((type) => (
                    <Chip key={type}>{type}</Chip>
                  ))}
                </div>

                {/* Image */}
                {/* eslint-disable-next-line @next/next/no-img-element */}
                <img
                  style={{ maxHeight: '120px' }}
                  className="ml-auto rounded-lg border-slate-300 border p-2 bg-white"
                  src={pokemon.image}
                  aria-label={pokemon.name}
                  alt={pokemon.name}
                />
              </div>
            </div>
          </motion.div>
        </Link>
      ))}
    </div>
  );
};
