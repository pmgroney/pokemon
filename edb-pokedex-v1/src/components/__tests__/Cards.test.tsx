import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import { Cards } from '@/components/Cards';

describe('Cards', () => {
  it('should render list of pokemons', async () => {
    const data = [
      {
        id: 1,
        name: 'Bulbasaur',
        image: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
        stats: {
          HP: 45,
          Attack: 49,
          Defense: 49,
          'Sp. Attack': 65,
          'Sp. Defense': 65,
          Speed: 45,
        },
      },
      {
        id: 2,
        name: 'Ivysaur',
        image: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
        stats: {
          HP: 60,
          Attack: 62,
          Defense: 63,
          'Sp. Attack': 80,
          'Sp. Defense': 80,
          Speed: 60,
        },
      },
    ];

    const { container } = render(<Cards data={data} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByTestId('pokemon-1-name')).toHaveTextContent('Bulbasaur');
  });
});
