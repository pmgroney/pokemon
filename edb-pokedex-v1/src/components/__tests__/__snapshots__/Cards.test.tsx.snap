// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards should render list of pokemons 1`] = `
<div>
  <div
    class="flex flex-wrap gap-10 justify-between"
  >
    <div
      class="rounded-lg bg-slate-100 border-slate-300 border p-5 max-w-md cursor-pointer flex-none w-6/12"
    >
      <div>
        <div
          class="text-2xl"
          data-testid="pokemon-1-name"
        >
          Bulbasaur
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="flex gap-1 flex-col items-start"
          />
          <img
            alt="Bulbasaur"
            aria-label="Bulbasaur"
            class="ml-auto rounded-lg border-slate-300 border p-2 bg-white"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
            style="max-height: 120px;"
          />
        </div>
      </div>
    </div>
    <div
      class="rounded-lg bg-slate-100 border-slate-300 border p-5 max-w-md cursor-pointer flex-none w-6/12"
    >
      <div>
        <div
          class="text-2xl"
          data-testid="pokemon-2-name"
        >
          Ivysaur
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="flex gap-1 flex-col items-start"
          />
          <img
            alt="Ivysaur"
            aria-label="Ivysaur"
            class="ml-auto rounded-lg border-slate-300 border p-2 bg-white"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
            style="max-height: 120px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
